### DISCUSSION TITLE
Can't download videos from facebook

### DISCUSSION THREAD

**Discussion by KillerBoogie** [aclap-dev/video-downloadhelper#249] [Help and Q&A]
### Your question

For while I can't download videos from facebook. I'm just seeing the option to downlaod a bunch of small mp4 files in size between a few KB or MB. Here is an example: https://www.facebook.com/kelly.arsenault.7524/videos/3535670640080086/

Downloads from YouTube work.

### Addon & CoApp version and details

version: 8.2.1.1
channel: stable
build date: 2024-02-19
build options: linuxlic: false, noyt: false, browser: firefox
lang: de
license: accepted
key (only 16 first characters): aa4dc7eb53874527
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.19",
  "binary": "C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "windows",
    "arch": "x86_64",
    "node": "18"
  },
  "home": "C:\\Users\\micro",
  "converterBinary": "C:\\Program Files\\DownloadHelper CoApp\\ffmpeg.exe",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "c30f360955"
}

**Comment by Mattmjs** [comment]
I don't use Facebook myself but have you tried the mobile trick replacing **www** with **m** in the facebook link.
The only problem with that is it provides the lowest quality and you may be after the highest quality.
From what I've observed they don't appear to use any streaming manifests.

@paulrouget , if you're reading this may be helpful. This may be how Facebook handles streaming media.


![Screenshot 2024-03-04 213408](https://github.com/aclap-dev/video-downloadhelper/assets/160297904/57cf3d59-d8d3-42ec-a566-7e72e8d8c238)

**Reply by KillerBoogie** [reply]
Using "m." as the host name doesn't work for me. Facebook also autoamtically switches to www. 
I had that issue before and a reset fixed it. I tried to reset teh settings to standard, but it still doesn't work.

**Reply by jcfrog** [reply]
Hi,
This should help: https://its.uri.edu/services/946442a2ab7e3cfcb0d9034469b32186c7fad400e8/simulating-a-mobile-environment-in-a-browser/
You need to specify a mobile user agent to avoid www redirection

**Reply by Mattmjs** [reply]
@jcfrog 
That's something I didn't know about before. Learned something new.

@KillerBoogie 
Surprised Facebook won't allow the simple change in the url. I am able to do it but I'm not a member of Facebook so that means
not logged / signed into the site.

**Comment by Phatman113** [comment]
I also have this issue, but I'm looking to download the highest quality if i can.  Anyone been able to correct that?  Trying the dev tools let me set mobile, but when I select a device with a large screen size, it starts erroring on download again...

**Reply by Mattmjs** [reply]
The mobile trick won't provide the highest quality, you could fish out the best quality from the developer tools.
The video and audio are separate files and this might be too much of a hassle.

**Comment by GrampaWildWilly** [comment]
I am running Windows 11 Pro 64-bit version 23H2 build 22631.3235, Firefox 123.0.1 64-bit, licensed VDH 8.2.2.6a1 beta, CoApp 2.0.19.  That is pretty much irrelevant because VDH does not work on the Facebook page mentioned in the original post in this discussion.  I do not work for VDH.  I am just another user, like you.

Now this is probably more trouble than it's worth for most folks.  But I'm going to show you how to do this on the off chance that somebody is actually sufficiently motivated to apply this approach.  So here goes.

As preparation for what follows, I performed these steps in this order.  The order is important.

1. Copy the Facebook URL from the first post in this thread to my system clipboard.
2. Open a new browser window.
3. Open the Network Monitor.  This is F12 in Firefox on Windows.  For other browsers on other platforms, I leave it to you to figure out how to do it on your system.
4. Paste the URL from step 1 into the URL bar at the top of the browser window.
5. Hit Enter.

The Network Monitor was full of much stuff that I couldn't use.  I am skipping several minutes of trial & error here & cutting directly to what worked.  I discovered by persistence & guessing that this was useful:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/b3a7c9d2-9610-48d5-8d68-17e87c12d5f8)

These images may appear completely illegible to you because they are small.  Just click the image.  GitHub will open the image in a new browser window.  In there, you can use the browser's magnifier (Ctrl-+) to make the image easily legible.

What you see there is the popup context menu on the highlighted entry in the Network Monitor.  When I executed the Save Response As menu option, I got this file selection dialog:

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/f6c714af-f275-40c1-9fc6-aa06a7c6350f)

That file name has no extension.  I added .txt.

![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/86e90496-8de3-47c1-8e62-d3e0012819e7)

After that saved the file on my system, I opened it in Notepad++, a Notepad replacement.  The file is just plain text.   You can open it in Notepad or any other text editor.  Don't use a word processor or anything else fancier than a text editor.  There are many nice features of Notepad++ that make this somewhat easier to deal with.  You might notice some as we go along here.

So after I opened the saved file, I did a search for the string mimetype:

![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/5c378697-2d7e-4c60-bfc3-45adf3bc5511)

After I dismissed the find dialog, I scrolled the page down using the mouse.  I normally avoid the mouse as much as I can, but in this case, scrolling with the keyboard would lose the highlighting you see in this next image.

![#05](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/41ec26e8-f3fe-4bf6-a1d6-09bb4a7a5066)

There are several mimetype descriptors in this file.  Let's inspect the first one closely.  I have highlighted the first mimetype descriptor in this next image:

![#05a](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/a07208e7-0a6f-4351-9196-9e8ccc660e7c)

To help me explain what is in this descriptor, I copy/pasted that monstrosity out to its own window:

![#06](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/d72fa0d8-abd3-44b4-aff7-32af0f030743)

I then inserted line breaks so I could highlight the parts that are important.

![#07](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/e4ea19d0-c3a4-4fd9-8417-34fe1b4f519a)

The first highlight tells us that this describes a video/mp4.

The second highlight tells us that this video has resolution 640x360.

The third highlight is the URL of this video.  But this is not yet a URL that we can use.  It needs to be massaged a bit before we can use it.  Here, I'm showing what I copied out of the URL.  The URL is demarcated by the rather arcane string `\u003CBaseURL>` at the beginning & `\u003C\/BaseURL>` at the end.  I copied/pasted just the URL from the middle of that out to its own window.

![#08](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/afada43c-5c80-447b-84c2-4510f59b1ba0)

In that new window that contains only the URL, I made 2 modifications.  First, I removed all the back slashes.

![#09](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/d936917f-e277-48dc-aa89-8279991a931e)
![#10](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/1115ba80-6705-4cc4-978b-86199678dc6e)

There were 5 that needed to be removed.

I then replaced all the occurrences of `&amp;` with simple `&` symbols.

![#11](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/7af10188-2c8a-4c71-9f45-c4ad692dfffe)

There were 7 of those.

At this point, I had a usable URL.  I could just double click it in the Notepad++ window & that would launch that URL in its own browser window.  If you don't have Notepad++, you can just copy/paste the URL out of your text editor into a browser window.

But this is the URL for the rather putrid resolution 640x360.  We don't want that one.  We want the best one that is available here.  That happens to be 1920x1080.  There is a mimetpye descriptor in here for 1920x1080.  Since I am here teaching, I am going to do what any good teacher would do.  I am setting a test for you to prove that you have been paying attention.  Find the 1920x1080 mimetype descriptor.  Come on now.  Don't give me any of that "I'm not very technical" junk.  What have I done here?  I've done some find & replace operations in a text editor.  Don't tell me using a text editor is beyond your technical abilities.  Find the 1920x1080 mimetype descriptor.  Then find the audio/mp4 mimetype descriptor.  Be a good student and just do it.

Once you have those 2 URLs, you can open each one of them in its own browser tab.  Here I have opened the 1920x1080 video in its own browser tab:

![#12](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/8bf0300e-4f7e-47e7-a103-1adceabe012d)

Note the arrow at the bottom.  That shows that this is a video track without an audio track.  I'm also showing a VDH menu.  Unlike a lot of cases we have encountered, this one is actually recognized by VDH.  I picked the first variant off the VDH menu & downloaded it.  I don't know what the second variant is.  VDH sorts its menu in descending order of what you should prefer to download.  It always does that.  So I ignored the second variant on the VDH menu.  I did the obvious thing & downloaded the first variant.

I also opened the audio track in its own browser tab:

![#13](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/e0097359-8506-4455-b655-d3599960e3c6)

This isn't showing the first frame of video but the little speaker icon on the player doesn't have a line through it.  So this is an audio track without a video track.  Once again, we're in luck & VDH can download this.  Once again, I downloaded the first variant on the VDH menu.

Here are the results:

![#14](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/74021728-e7d3-4492-83a8-d167df3c6fa0)

The video downloaded in 10 seconds & the audio took less than 1 second.  Just calculate the modified time minus the creation time.  Note that there is video information but no audio information for the one file, & audio information but no video information for the other file.  This is what we were expecting.  Also note that the video has resolution 1920x1080.

VLC has a synchronous playback feature.  You get it from the Open Multiple Files selection, Ctrl+Shift+o.  (That's letter o, not zero.)  I just played the 2 files as they were:

![#15](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/6ad41a52-a283-4774-a814-659a284c0e8b)

You can see it was playing fine.  You'll have to trust me that the audio was equally fine.  From beginning to end.  No problems.

If you absolutely must have this as a unified video file, you can merge the 2 tracks using VDH's tool for that purpose.

Like I said, it's a bit of trouble.  Not really complicated, just a bunch of easy steps.  Anybody can do this.  No programming is involved.  This doesn't require technical expertise.  It's up to you whether you absolutely have to have Facebook videos.  Just apply this approach & you can have them all.  Well, that might be an extravagant claim.  I'm not a member of Facebook so I'm probably extrapolating a trend from a single data point when I say that.  But I would imagine that Facebook does not do things differently for each video on their site.  That would be insanity.  I have to assume that they do the same thing with all their videos.  So you can get things to work by taking this approach.  That seems to be your only recourse unless & until @paulrouget decides to tackle this issue & provide a simple solution in VDH.  You'll note that I did use VDH here to download this content.  I think that means that VDH can probably do under program control what I did here manually.  But until that becomes available, just follow my instructions & you'll be in fat city.

**Reply by GrampaWildWilly** [reply]
I have been noticing that sometimes, I click on an image, GitHub opens a new browser window, & the window remains blank.  I don't know what's up with this.  It seems to be a quirk of GitHub.  Just reload this web page & try again.  That seems to make things work properly . . . until the next time they don't.  It is a recurring issue.  Kind of annoying, but hey, this seems to be the reality of things.

**Reply by Mattmjs** [reply]
That's a lot of work to get a higher quality video. But I hadn't seen Save Response As before, must have been recently added.
In my screen shot above there is an entry I highlighted which has streaming information in it.
It will only appear on mousing over the video like what is done on YouTube. I did that from **facebook.com/watch**
If that entry isn't found then you've shown the same information is in the html document.

**Reply by GrampaWildWilly** [reply]
I had a vague feeling that context menu has changed.  I think a couple of things aren't in positions they used to be & there's a couple of new functions.  As for the alleged html document, what I saved didn't look anything like actual HTML.  Not that I'm an expert on HTML, but it didn't look like HTML.  It looked more like what we've gotten from json objects in the past.  It kinda sorta looked like XML.  We're kind of reverse engineering the web pages to find where the media information is hiding.  I'm just looking for anything that gives us m3u8 objects or strings that look like URLs.  Sometimes I get lucky, sometimes I strike out.  That's just evidence that I don't really know what I'm doing.  I just flounder around until something works.
